tools/interop_matrix/run_interop_matrix_tests.py

```diff
@@ -92,136 +93,149 @@
-                    if args.server_host:
-                        if line.find('--server_host=') > -1:
-                            line = re.sub('--server_host=[^ ]*',
-                                          '--server_host=%s' % args.server_host,
-                                          line)
-                        else:
-                            line = '%s --server_host=%s"' % (line[:-1],
-                                                             args.server_host)
-                        print(line)
-
-                    spec = jobset.JobSpec(
-                        cmdline=line,
-                        shortname='%s:%s:%s:%s' % (suite_name, lang, server,
-                                                   shortname),
-                        timeout_seconds=_TEST_TIMEOUT,
-                        shell=True,
-                        flake_retries=5 if args.allow_flakes else 0)
-                    job_spec_list.append(spec)
-            jobset.message(
-                'START',
-                'Loaded %s tests from %s' % (len(job_spec_list), testcases),
-                do_newline=True)
-    except IOError as err:
-        jobset.message('FAILED', err, do_newline=True)
+    for line in testcase_lines:
+        m = re.search('--test_case=(.*)"', line)
+        shortname = m.group(1) if m else 'unknown_test'
+        m = re.search('--server_host_override=(.*).sandbox.googleapis.com',
+                      line)
+        server = m.group(1) if m else 'unknown_server'
+
+        # If server_host arg is not None, replace the original
+        # server_host with the one provided or append to the end of
+        # the command if server_host does not appear originally.
+        if args.server_host:
+            if line.find('--server_host=') > -1:
+                line = re.sub('--server_host=[^ ]*',
+                              '--server_host=%s' % args.server_host, line)
+            else:
+                line = '%s --server_host=%s"' % (line[:-1], args.server_host)
+
+        spec = jobset.JobSpec(
+            cmdline=line,
+            shortname='%s:%s:%s:%s' % (suite_name, lang, server, shortname),
+            timeout_seconds=_TEST_TIMEOUT_SECONDS,
+            shell=True,
+            flake_retries=5 if args.allow_flakes else 0)
+        job_spec_list.append(spec)
     return job_spec_list
 
 
-_xml_report_tree = report_utils.new_junit_xml_tree()
+def _pull_images_for_lang(lang, images):
+    """Pull all images for given lang from container registry."""
+    jobset.message(
+        'START', 'Downloading images for language "%s"' % lang, do_newline=True)
+    download_specs = []
+    for release, image in images:
+        # Pull the image and warm it up.
+        # First time we use an image with "docker run", it takes time to unpack the image
+        # and later this delay would fail our test cases.
+        cmdline = [
+            'gcloud docker -- pull %s && docker run --rm=true %s /bin/true' %
```

Comment:
I didn't know that either, but this seems to be the reason why the parallel downloads had to be rolled back in  #15701 (I realized from the logs that the first time a test is run by `docker run` after pulling the image, it takes much longer to run than the subsequent uses of the same image (this was actually causing the first test case to time out).

Reply:
Can we be certain that this is strictly a `docker run` issue and not another kind of cache-miss-then-hit-on-subsequent-test-runs that's outside of docker and possibly due to grpc clients? There are other places that `docker run` on a freshly pulled image without a warmup `docker run` first - albeit maybe not with a short timeout on the `docker run`. Otherwise this seems like a bad docker gotcha